File: Java/WebsiteCode/LeechCode/src/Medium/Array/FindIfArrayCanBeSorted.java
package Medium.Array;

import java.util.Arrays;

public class FindIfArrayCanBeSorted {
    public static void main(String[] args) {
        int[][] tests = {
            {8,4,2,30,15}
        };

        for (int[] nums : tests) {
            System.out.println(new FindIfArrayCanBeSorted_Solution().canSortArray(nums));
        }
    }
}

// 5ms 44.67MB
class FindIfArrayCanBeSorted_Solution {

    public boolean canSortArray(int[] nums) {
        int[] values = Arrays.copyOf(nums, nums.length);

        int n = values.length;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (values[j] <= values[j + 1]) {
                    continue;
                } else {
                    if (
                        Integer.bitCount(values[j]) ==
                        Integer.bitCount(values[j + 1])
                    ) {
                        int temp = values[j];
                        values[j] = values[j + 1];
                        values[j + 1] = temp;
                    } else {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}

File: PHP/WebsiteCode/LeechCode/Medium/Array/FindIfArrayCanBeSorted.php
<?php 

// 9ms 19.4MB
class Solution2
{
    /**
     * @param int[] $nums
     * @return bool
     */
    function canSortArray($nums)
    {
        $currentMin = 1024;
        $currentMax = 0;
        $prevGroupMax = null;
        foreach($nums as $i => $num) {
            if (($i > 0 && !$this->isSameSetBits($nums[$i-1], $num))) {
                if ($prevGroupMax > $currentMin) return false;
                $prevGroupMax = $currentMax;
                $currentMin = 1024;
                $currentMax = 0;
            }
            $currentMin = min($currentMin, $num);
            $currentMax = max($currentMax, $num);
        }

        if ($prevGroupMax > $currentMin) return false;

        return true;
    }

    function isSameSetBits($a, $b)
    {
        return $this->countBits($a) === $this->countBits($b);
    }

    function countBits($a)
    {
        $countA = 0;
        while ($a > 0) {
            $countA += $a % 2;
            $a = $a >> 1;
        }
        return $countA;
    }
}
// 111ms 19.70MB
class Solution {
    
    /**
     * @param integer[] $nums
     * @return bool
     */
    function canSortArray($nums):bool {
        $len = count($nums);
        
        for ($i = 0; $i < $len; $i++) {
            for ($j = 0; $j < $i; $j++) {
                if ($nums[$i] < $nums[$j]) {
                    if (
                        substr_count(decbin($nums[$i]),'1') ==
                        substr_count(decbin($nums[$j]),'1')
                    ) {
                        // Swap the elements
                        $temp = $nums[$i];
                        $nums[$i] = $nums[$j];
                        $nums[$j] = $temp;
                    } else {
                        return false;
                    }
                }
            }
        }

        var_dump($nums);
        return true;
    }
}